File: typescript/src/code-smells/oop-abusers/switch-statements.ts
export type EmployeeKind = 'engineer' | 'manager' | 'sales'

export interface EmployeeRecord {
  kind: EmployeeKind
  base: number
  bonus?: number
  commission?: number
}

export function calculatePay(rec: EmployeeRecord): number {
  switch (rec.kind) {
    case 'engineer':
      return rec.base
    case 'manager':
      return rec.base + (rec.bonus ?? 0)
    case 'sales':
      return rec.base + (rec.commission ?? 0)
    default:
      const _exhaustive: never = rec.kind
      return _exhaustive
  }
}

export function demoSwitchStatements(): number[] {
  return [
    calculatePay({kind: 'engineer', base: 1000}),
    calculatePay({kind: 'manager', base: 1000, bonus: 200}),
    calculatePay({kind: 'sales', base: 800, commission: 500}),
  ]
}

File: csharp/src/code-smells/oop-abusers/SwitchStatements.cs
namespace RefactoringAvanzado.CodeSmells.OopAbusers;

public enum EmployeeKind
{
    Engineer,
    Manager,
    Sales
}

public record EmployeeRecord(EmployeeKind Kind, double Base, double? Bonus = null, double? Commission = null);

public static class SwitchStatements
{
    public static double CalculatePay(EmployeeRecord rec)
    {
        return rec.Kind switch
        {
            EmployeeKind.Engineer => rec.Base,
            EmployeeKind.Manager => rec.Base + (rec.Bonus ?? 0),
            EmployeeKind.Sales => rec.Base + (rec.Commission ?? 0),
            _ => throw new ArgumentException($"Unknown employee kind: {rec.Kind}")
        };
    }

    public static double[] DemoSwitchStatements()
    {
        return
        [
            CalculatePay(new EmployeeRecord(EmployeeKind.Engineer, 1000)),
            CalculatePay(new EmployeeRecord(EmployeeKind.Manager, 1000, Bonus: 200)),
            CalculatePay(new EmployeeRecord(EmployeeKind.Sales, 800, Commission: 500))
        ];
    }
}
